Source/JavaScriptCore/wasm/js/WebAssemblyModuleConstructor.cpp

```diff
@@ -101,6 +101,104 @@ JSC_DEFINE_HOST_FUNCTION(webAssemblyModuleCustomSections, (JSGlobalObject* globa
     return JSValue::encode(result);
 }
 
+template<typename T>
+concept IsImportOrExport = std::same_as<T, Wasm::Import> || std::same_as<T, Wasm::Export>;
+
+// https://github.com/WebAssembly/js-types
+// https://webassembly.github.io/js-types/js-api/index.html#dictdef-anyexterntype
+template <IsImportOrExport T>
+static JSObject* createTypeReflectionObject(JSGlobalObject* globalObject, JSWebAssemblyModule* module, const T& impOrExp)
+{
+    VM& vm = globalObject->vm();
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+
+    JSObject* typeObj;
+    constexpr auto errorMessage = std::same_as<T, Wasm::Import> ? "WebAssembly.Module.imports unable to produce import descriptors for the given module"_s : "WebAssembly.Module.exports unable to produce export descriptors for the given module"_s;
+    switch (impOrExp.kind) {
+    case Wasm::ExternalKind::Function: {
+        typeObj = constructEmptyObject(globalObject, globalObject->objectPrototype(), 2);
+
+        Wasm::TypeIndex typeIndex = module->moduleInformation().typeIndexFromFunctionIndexSpace(impOrExp.kindIndex);
+        const auto& signature = Wasm::TypeInformation::getFunctionSignature(typeIndex);
+
+        JSArray* functionParametersTypes = constructEmptyArray(globalObject, nullptr);
+        RETURN_IF_EXCEPTION(throwScope, { });
+        auto argumentCount = signature.argumentCount();
+        for (unsigned i = 0; i < argumentCount; ++i) {
+            JSString* typeString = Wasm::typeToJSAPIString(vm, signature.argumentType(i));
+            if (!typeString) {
+                throwException(globalObject, throwScope, createTypeError(globalObject, errorMessage));
+                return nullptr;
+            }
```

Comment:
Such `!typeString` checks do the same thing as https://commits.webkit.org/274900@main